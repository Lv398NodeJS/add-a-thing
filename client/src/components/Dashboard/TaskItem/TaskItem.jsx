import React, { Component } from 'react';
import {
  Container,
  Row,
  Button,
  Col,
} from 'react-bootstrap';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import TaskDetailsModal from '@TaskDetails/TaskDetailsModal';
import * as taskActions from '@actions/taskDetailsActions';
import * as subtaskActions from '@actions/subtaskActions';
import './TaskItem.scss';
import del from '@assets/delete.svg';
import accept from '@assets/accept.svg';
import {
  getTaskStyleByPriority,
  getTaskStyleByStatus,
  dragStart,
  dragEnd,
} from './TaskItemUtils';

class TaskItem extends Component {
  constructor() {
    super();
    this.state = {
      modalShow: false,
      isDeleted: false,
    };
  }

  openTaskDetails = () => {
    const {
      id,
      modalOpen,
      taskDetailsActions: { fetchTaskDetails },
      subtaskListActions: { fetchSubtaskList },
    } = this.props;
    fetchTaskDetails(id);
    fetchSubtaskList(id);
    this.setState({ modalShow: !modalOpen });
  }

  closeTaskDetails = () => {
    this.setState({ modalShow: false });
  };

  deleteTaskHandle = (event) => {
    const { isDeleted } = this.state;
    const { taskDetailsActions: { deleteTaskDetails }, id } = this.props;

    if (!isDeleted) {
      this.setState({ isDeleted: true });
    } else {
      deleteTaskDetails(id);
    }

    event.stopPropagation();
  }

  render() {
    const {
      id,
      taskName,
      status,
      priority,
    } = this.props;

    const {
      isDeleted,
      modalShow: modalOpen,
    } = this.state;

    return (
      <Container
        className="p-0"
        fluid="true"
      >
        <Container
          id={id}
          fluid="false"
          data-test="taskContainer"
          draggable="true"
          className={getTaskStyleByPriority(priority)}
          onClick={this.openTaskDetails}
          onDragStart={e => dragStart(e)}
          onDragEnd={e => dragEnd(e)}
        >
          <Col
            lg={11}
            md={10}
          >
            <Row>
              <Container
                className={getTaskStyleByStatus(status)}
                data-test="taskName"
              >
                <span>
                  {taskName}
                </span>
              </Container>
            </Row>
          </Col>
          <Col
            lg={1}
            md={2}
            className="d-flex justify-content-end align-items-center m-0 p-0"
          >
            <Button
              variant="light"
              className="delete-button"
              size="sm"
              onClick={this.deleteTaskHandle}
              onMouseLeave={() => this.setState({ isDeleted: false })}
            >
              {!isDeleted
                ? <img src={del} alt="Delete" className="delete-icon" draggable="false" />
                : <img src={accept} alt="Accept" className="accept-icon" draggable="false" />}
            </Button>
          </Col>
        </Container>
        <Container>
          <TaskDetailsModal
            data-test="taskDetails"
            show={modalOpen}
            onClose={() => { this.closeTaskDetails(); }}
          />
        </Container>
      </Container>
    );
  }
}

const mapStateToProps = ({ mainContainerReducer: { taskListRef } }) => ({
  taskListRef,
});

const mapDispatchToProps = dispatch => ({
  taskDetailsActions: bindActionCreators(taskActions, dispatch),
  subtaskListActions: bindActionCreators(subtaskActions, dispatch),
});

export { TaskItem as TaskItemComponent };
export default connect(mapStateToProps, mapDispatchToProps)(TaskItem);
