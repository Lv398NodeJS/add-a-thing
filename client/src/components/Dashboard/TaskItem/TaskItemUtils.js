export const getTaskStyleByPriority = (priority) => {
  const priorityClass = `${priority}-task`;
  const finalStyle = `task-item d-flex justify-content-start ${priority && priorityClass}`;
  return finalStyle;
};

export const getTaskStyleByStatus = (status) => {
  const style = status === 'Done' ? 'task-item-crossed' : '';
  return style;
};

export const dragStart = (event) => {
  event.dataTransfer.setData('taskID', event.target.id);
  event.target.classList.add('dragged', 'dragged-task');

  const fakeTask = event.target.cloneNode(true);
  fakeTask.setAttribute('id', 'drag-avatar');
  event.target.parentNode.parentNode.appendChild(fakeTask);

  event.dataTransfer.setDragImage(fakeTask, 60, 25);
  document.getElementById('delete-zone').classList.add('shown');
};

export const dragEnd = (event) => {
  const fakeTask = document.getElementById('drag-avatar');
  const REMOVE_DELAY = 80;
  if (fakeTask != null) fakeTask.remove();
  event.target.classList.remove('dragged');
  setTimeout(() => document.getElementById('delete-zone').classList.remove('shown'), REMOVE_DELAY);
};
