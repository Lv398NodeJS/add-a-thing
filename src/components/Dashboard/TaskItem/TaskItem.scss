$task-item-text: #757477;
$task-item-background: linear-gradient(to right, #ececec, #f6f6f6);
$high-priority: red;
$medium-priority: orange;
$low-priority: green;

.task-item,
.dragged {
  padding: 7px;
  margin: 18px auto;
  min-height: 38px;
  border-radius: 2px;
  font-size: 1rem;
  background: $task-item-background;
  cursor: grab;
  overflow: hidden;
};

.task-item span {
  text-align: left;
  hyphens: auto;
  overflow-wrap: break-word;
  word-break: break-word;
}

.dragged {
  opacity: 0;
}

#drag-avatar {
  outline: 1px dashed $task-item-text;
  opacity: 1;
  transform: translateX(-200%);
}

.task-item-crossed {
  text-decoration: line-through;
}

.task-item-container {
  padding: 0;
}

.High-task {
  border-left: 2px solid $high-priority;
}

.Medium-task {
  border-left: 2px solid $medium-priority;
}

.Low-task {
  border-left: 2px solid $low-priority;
}

.delete-button {
  &, &:hover, &:link, &:focus, &:active {
    -webkit-box-shadow: none !important;
    -moz-box-shadow: none !important;
    box-shadow: none !important;
    background: transparent !important;
    border: none;
    outline: none;
    box-shadow: none;
    padding-top: 0;
    margin: -10px 0;
    opacity: 0;
  }
  &:hover {
    opacity: 1;
    transform: scale(1.5);
    transition: transform 150ms ease;
  }
  &:active {
    transform: scale(1)
  }
}

.delete-icon {
  width: 1rem;
}

.accept-icon {
  width: 0.8rem;
  margin-left: 0.1rem;
}