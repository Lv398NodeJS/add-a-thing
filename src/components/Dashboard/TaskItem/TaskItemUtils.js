export const getTaskRef = (taskListRef, id) => {
  const taskRef = taskListRef && taskListRef.child(`${id}`);
  return taskRef;
};

export const getTaskStyleByPriority = (priority) => {
  const priorityClass = `${priority}-task`;
  const finalStyle = `task-item d-flex justify-content-between ${priority && priorityClass}`;
  return finalStyle;
};

export const getTaskStyleByStatus = (status) => {
  const style = status === 'Done' ? 'task-item-crossed' : '';
  return style;
};

export const dragStart = (event) => {
  event.dataTransfer.setData('taskID', event.target.id);
  event.target.classList.add('dragged');

  const fakeTask = event.target.cloneNode(true);
  fakeTask.setAttribute('id', 'drag-avatar');
  event.target.parentNode.parentNode.appendChild(fakeTask);

  event.dataTransfer.setDragImage(fakeTask, 60, 25);
  document.getElementById('delete-zone').classList.add('shown');
};

export const dragEnd = (event) => {
  event.target.classList.remove('dragged');

  const fakeTask = document.getElementById('drag-avatar');
  if (fakeTask != null) fakeTask.remove();

  setTimeout(() => document.getElementById('delete-zone').classList.remove('shown'), 80);
};
