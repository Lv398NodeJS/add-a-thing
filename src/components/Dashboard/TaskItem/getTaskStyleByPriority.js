export default (priority, status) => {
  let style = status === 'Done' ? 'taskItemDone' : 'taskItem';
  switch (priority) {
    case 'High':
      style += ' priorityH';
      break;
    case 'Medium':
      style += ' priorityM';
      break;
    case 'Low':
      style += ' priorityL';
      break;
    default:
      break;
  }
  return style;
};
