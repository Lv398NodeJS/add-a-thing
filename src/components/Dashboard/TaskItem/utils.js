export const getTaskRef = (props) => {
  const { taskListRef, id } = props;
  const taskRef = taskListRef.child(`${id}`);
  return taskRef;
};

export const getTaskStyleByPriority = (props) => {
  const { priority, status } = props;
  const lineStyle = status === 'Done' ? 'taskItemDone' : 'taskItem';
  const style = `${lineStyle} ${priority}`;
  return style;
};
