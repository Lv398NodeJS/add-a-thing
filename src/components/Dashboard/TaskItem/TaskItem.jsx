import React, { Component } from 'react';
import { Container } from 'react-bootstrap';
import './TaskItem.scss';
import getTaskStyleByPriority from './getTaskStyleByPriority';
import TaskDetails from '../../TaskDetails/TaskDetails';

export default class TaskItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      modalShow: false,
    };
  }

  closeTaskDetails = () => {
    this.setState({ modalShow: false });
  };

  render() {
    const {
      taskName, taskListRef, id, priority, status,
    } = this.props;
    const taskRef = taskListRef.child(`${id}`);
    const { modalShow: modalOpen } = this.state;

    return (
      <Container className="TaskItemContainer">
        <Container
          className={getTaskStyleByPriority(priority, status)}
          onClick={() => this.setState({ modalShow: !modalOpen })}
        >
          {taskName}
        </Container>
        <Container>
          <TaskDetails
            taskRef={taskRef}
            show={modalOpen}
            onClose={() => { this.closeTaskDetails(); }}
          />
        </Container>
      </Container>
    );
  }
}
