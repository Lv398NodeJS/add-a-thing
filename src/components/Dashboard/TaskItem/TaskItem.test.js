import React, { Container } from 'react';
import { shallow } from 'enzyme';
import TaskItem from './TaskItem';


describe('TaskItem displays correctly', () => {
  it('should render task name', () => {
    const wrapper = shallow(<TaskItem />);
    const taskName = wrapper.find(<Container id={10} />);
    expect(taskName).to.be.equal(3);
  });
});
