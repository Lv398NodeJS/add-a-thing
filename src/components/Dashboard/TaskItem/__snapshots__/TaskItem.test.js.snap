// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskItem /> should match snapshot when render name from props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TaskItem
    taskName="Task 1"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "as": "div",
      "children": Array [
        <ForwardRef
          as="div"
          data-test="taskName"
          fluid={true}
          onClick={[Function]}
        >
          Task 1
        </ForwardRef>,
        <ForwardRef
          as="div"
          fluid={false}
        >
          <TaskDetailsModal
            data-test="taskDetails"
            onClose={[Function]}
            show={false}
            taskRef={20}
          />
        </ForwardRef>,
      ],
      "className": "task-item-container",
      "fluid": true,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "div",
          "children": "Task 1",
          "className": undefined,
          "data-test": "taskName",
          "fluid": true,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Task 1",
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "as": "div",
            "fluid": false,
          },
          "displayName": "Container",
          "render": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "div",
          "children": <TaskDetailsModal
            data-test="taskDetails"
            onClose={[Function]}
            show={false}
            taskRef={20}
          />,
          "fluid": false,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "data-test": "taskDetails",
            "onClose": [Function],
            "show": false,
            "taskRef": 20,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "as": "div",
            "fluid": false,
          },
          "displayName": "Container",
          "render": [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": Object {
        "as": "div",
        "fluid": false,
      },
      "displayName": "Container",
      "render": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "as": "div",
        "children": Array [
          <ForwardRef
            as="div"
            data-test="taskName"
            fluid={true}
            onClick={[Function]}
          >
            Task 1
          </ForwardRef>,
          <ForwardRef
            as="div"
            fluid={false}
          >
            <TaskDetailsModal
              data-test="taskDetails"
              onClose={[Function]}
              show={false}
              taskRef={20}
            />
          </ForwardRef>,
        ],
        "className": "task-item-container",
        "fluid": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "div",
            "children": "Task 1",
            "className": undefined,
            "data-test": "taskName",
            "fluid": true,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Task 1",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "as": "div",
              "fluid": false,
            },
            "displayName": "Container",
            "render": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "div",
            "children": <TaskDetailsModal
              data-test="taskDetails"
              onClose={[Function]}
              show={false}
              taskRef={20}
            />,
            "fluid": false,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "data-test": "taskDetails",
              "onClose": [Function],
              "show": false,
              "taskRef": 20,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "as": "div",
              "fluid": false,
            },
            "displayName": "Container",
            "render": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "as": "div",
          "fluid": false,
        },
        "displayName": "Container",
        "render": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`<TaskItem /> should render and match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TaskItem />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "as": "div",
      "children": Array [
        <ForwardRef
          as="div"
          data-test="taskName"
          fluid={true}
          onClick={[Function]}
        />,
        <ForwardRef
          as="div"
          fluid={false}
        >
          <TaskDetailsModal
            data-test="taskDetails"
            onClose={[Function]}
            show={false}
          />
        </ForwardRef>,
      ],
      "className": "task-item-container",
      "fluid": true,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "div",
          "children": undefined,
          "className": undefined,
          "data-test": "taskName",
          "fluid": true,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "as": "div",
            "fluid": false,
          },
          "displayName": "Container",
          "render": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "div",
          "children": <TaskDetailsModal
            data-test="taskDetails"
            onClose={[Function]}
            show={false}
          />,
          "fluid": false,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "data-test": "taskDetails",
            "onClose": [Function],
            "show": false,
            "taskRef": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "as": "div",
            "fluid": false,
          },
          "displayName": "Container",
          "render": [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": Object {
        "as": "div",
        "fluid": false,
      },
      "displayName": "Container",
      "render": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "as": "div",
        "children": Array [
          <ForwardRef
            as="div"
            data-test="taskName"
            fluid={true}
            onClick={[Function]}
          />,
          <ForwardRef
            as="div"
            fluid={false}
          >
            <TaskDetailsModal
              data-test="taskDetails"
              onClose={[Function]}
              show={false}
            />
          </ForwardRef>,
        ],
        "className": "task-item-container",
        "fluid": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "div",
            "children": undefined,
            "className": undefined,
            "data-test": "taskName",
            "fluid": true,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "as": "div",
              "fluid": false,
            },
            "displayName": "Container",
            "render": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "div",
            "children": <TaskDetailsModal
              data-test="taskDetails"
              onClose={[Function]}
              show={false}
            />,
            "fluid": false,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "data-test": "taskDetails",
              "onClose": [Function],
              "show": false,
              "taskRef": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "as": "div",
              "fluid": false,
            },
            "displayName": "Container",
            "render": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "as": "div",
          "fluid": false,
        },
        "displayName": "Container",
        "render": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
